RISC-V: Update C calling conventions to use the ELF psABI

xavierleroy opened the pull request:
The original implementation of `loc_external_arguments` and `loc_external_results` was following an older ABI, where an FP argument passed in an FP register "burns" an integer register.

In the ELF psABI, integer registers and FP registers are used independently,as in the OCaml calling convention.

Fixes: #9515

rwmjones commented:
Does this change merit a regression test?

xavierleroy commented:
There's already a generic test for C calling conventions: https://github.com/ocaml/ocaml/tree/trunk/testsuite/tests/unboxed-primitive-args .  It's not obvious to me why it did not catch the issue earlier.  @jeremiedimino can you see why the case (unboxed float, untagged int) -> unboxed float was not exercised?

rwmjones commented:
Here's the latest build of OCaml on riscv64:

http://fedora.riscv.rocks/koji/buildinfo?buildID=156069

Going to Logs -> build.log shows ... oops, we disable the tests on riscv64, specifically here:

https://src.fedoraproject.org/rpms/ocaml/blob/master/f/ocaml.spec#_18

I guess the tests failed at some point in the past and someone (most likely me!) disabled them.  So that's not cool, I'll reenable the tests again when we update the compiler to 4.11 branch in the next few days.

nojb commented:
> There's already a generic test for C calling conventions: https://github.com/ocaml/ocaml/tree/trunk/testsuite/tests/unboxed-primitive-args . It's not obvious to me why it did not catch the issue earlier. @jeremiedimino can you see why the case (unboxed float, untagged int) -> unboxed float was not exercised?

Apparently the test is being skipped:
```
List of skipped tests:
    tests/unboxed-primitive-args/'test.ml' with 1.1.1 (ocaml) 
```
The test is also being skipped on other architectures it seems, such as Linux, but not BSDs. When trying to run the test individually, I get the following trace:
```
$ ../../../ocamltest/ocamltest.opt -e test.ml
Specified modules: test.ml
Source modules: test.ml
 ... testing 'test.ml' with 1 (hasunix) => Running test hasunix with 1 actions
Running action 1/1 (hasunix)
Action 1/1 (hasunix) passed (unix library available)
passed
 ... testing 'test.ml' with 1.1 (setup-ocamlopt.byte-build-env) => Running test setup-ocamlopt.byte-build-env with 1 actions
Running action 1/1 (setup-ocamlopt.byte-build-env)
Action 1/1 (setup-ocamlopt.byte-build-env) passed
passed
 ... testing 'test.ml' with 1.1.1 (ocaml) => Running test ocaml with 1 actions
Running action 1/1 (ocaml)
Action 1/1 (ocaml) skipped
skipped
 ... testing 'test.ml' with 1.1.1.1 (ocaml) => n/a
 ... testing 'test.ml' with 1.1.1.1.1 (ocamlopt.byte) => n/a
 ... testing 'test.ml' with 1.1.1.1.1.1 (run) => n/a
 ... testing 'test.ml' with 1.1.1.1.1.1.1 (check-program-output) => n/a
```
@shindere do you see why the `ocaml` action is being skipped?

nojb commented:
See #9517

shindere commented:
Thanks @nojb, you were faster than me!

nojb reviewed (changes requested):
According to the psABI, when there are more floating-point arguments than available floating-point registers, any remaining floating-point argument must be passed "according to the integer calling convention", which is to say that they should be passed using integer registers (if they are available), before using the stack. For example, the function
```c
double f (double a0, double a1, double a2, double a3,
          double a4, double a5, double a6, double a7, double a8)
{
  return a8;
}
```
is compiled to
```asm
f:
        addi    sp,sp,-96
        sd      s0,88(sp)
        addi    s0,sp,96
        fsd     fa0,-24(s0)
        fsd     fa1,-32(s0)
        fsd     fa2,-40(s0)
        fsd     fa3,-48(s0)
        fsd     fa4,-56(s0)
        fsd     fa5,-64(s0)
        fsd     fa6,-72(s0)
        fsd     fa7,-80(s0)
        sd      a0,-88(s0)
        fld     fa5,-88(s0)
        fmv.d   fa0,fa5
        ld      s0,88(sp)
        addi    sp,sp,96
        jr      ra
```
where the last floating-point argument is assumed to be passed in `a0`.

Unless I am missing something, this means that we need to insert conversion `float -> int` for any float argument after the 8th (if any) when calling external functions, right?

davidlt commented:
@jim-wilson would you have some feedback here?

xavierleroy commented:
@nojb is right about some FP arguments being passed in integer registers.  I forgot about this bit.  But the FP value must be preserved, so it's not float -> int conversions that need to be inserted, but FP reg -> int reg move instructions.  I'll come back with an implementation soon.

xavierleroy committed:
Support FP reg -> int reg moves

Using instruction fmv.x.d.

This is necessary to implement the ELF psABI calling conventions,
whereas some FP arguments may have to be passed in integer registers.

xavierleroy committed:
Update C calling conventions to the RISC-V ELF psABI

The original implementation of loc_external_arguments and
loc_external_results was following an older ABI,
where an FP argument passed in an FP register "burns" an integer register.

In the ELF psABI, integer registers and FP registers are used independently,
as in the OCaml calling convention.  Plus, if all FP registers are used
but an integer register remains, the integer register is used to pass
the next FP argument.

Fixes: #9515

xavierleroy commented:
Second try pushed to this branch.

nojb reviewed (approved):
LGTM (and it passes `unboxed-primitive-args`).

nojb commented:
Planning to squash-merge shortly unless someone objects.

xavierleroy commented:
Could you preserve the two commits please?  A non-squashed merge would be fine.

rwmjones commented:
No objection here.  We will be trying out this and related fixes when we do a new OCaml rebuild in a few days.

nojb commented:
Thank you everyone!

Octachron commented:
@rwmjones , I will publish an official alpha release for 4.11.0 in the beginning of next week, if you are interested with having a shared synchronization point.

rwmjones commented:
Yes that would be really useful.
